source/as-promise/utils/event-listeners-reorderer.ts

```diff
@@ -0,0 +1,60 @@
+export default class EventListenerReorderer {
```

Comment:
Is this really needed?

Reply:
I don't think it's really really needed, but I think is a lot better having that messy code (it can't be simplified too much) in a different file in order to not mess up `as-promise/index.ts`.

Reply:
I'm talking if we need to use `EventListenerReorderer` at all.

Reply:
Can't we just `request.socket.pause()` before the await and `request.socket.resume()` afterwards?

Reply:
> After the `error` event, the request fire the `response` event, but this is prevented by `request.destroy()`.
> The problem in my PR is that being the `error` listener async, the emitter sees it as immediately completed and goes on with the `response` event because `request.destroy()` has not been called yet.

That's the problem.

> The only thing that came to my mind was restoring the order of the two events (`error` and `response`) in the async environment.

An that's a solution I found.

If you've a better solution please let me know, I really do not have any idea other than restoring the order of the two listeners.

Reply:
> Can't we just `request.socket.pause()` before the await and `request.socket.resume()` afterwards?

I'll give it a try

Reply:
> > Can't we just `request.socket.pause()` before the await and `request.socket.resume()` afterwards?
> 
> I'll give it a try

```
  2 tests failed
  1 known failure
  1 test skipped
  2 unhandled rejections
  3 uncaught exceptions
```

According to the Node.JS documentation
> socket.pause() Pauses the reading of data

The thing is that we need to "pause the events", when `error` is emitted I think all the data has been already read, so it not "paused" the response event.